apps/src/templates/admin/MassDeleteContainer.tsx

```diff
@@ -0,0 +1,438 @@
+            );
+          }
+
+          if (!headers.includes('unit_name')) {
+            throw new Error('CSV must have a column named "unit_name"');
+          }
+
+          // Process the data
+          const parsedData: OriginalData[] = results.data.map((row, index) => {
+            const studentId = row.student_id?.trim();
+            const studentUsername = row.student_username?.trim();
+            const unitName = row.unit_name?.trim();
+
+            if ((!studentId && !studentUsername) || !unitName) {
+              throw new Error(
+                `Row ${index + 2} has missing student identifier or unit_name`
+              );
+            }
+
+            const rowData: OriginalData = {
+              unit_name: unitName,
+            };
+
+            if (studentId) {
+              rowData.student_id = studentId;
+            }
+
+            if (studentUsername) {
+              rowData.student_username = studentUsername;
+            }
+
+            return rowData;
+          });
+
+          resolve(parsedData);
+        } catch (error) {
+          reject(error);
+        }
+      },
+      error: error => {
+        reject(new Error(`Failed to parse CSV: ${error.message}`));
+      },
+    });
+  });
+};
+
+const hasUsernameColumn = (data: OriginalData[]): boolean => {
+  return data.length > 0 && data[0].student_username !== undefined;
+};
+
+const getCSRFToken = (): string => {
+  const token = document
+    .querySelector('meta[name="csrf-token"]')
+    ?.getAttribute('content');
+  return token || '';
+};
+
+// Main container component
+const MassDeleteContainer: React.FC = () => {
+  const [csvFile, setCsvFile] = useState<File | null>(null);
+  const [csvData, setCsvData] = useState<Array<{
+    student_id?: string;
+    student_username?: string;
+    unit_name: string;
+  }> | null>(null);
+  const [csvError, setCsvError] = useState<string | null>(null);
+  const [hasUsernames, setHasUsernames] = useState<boolean>(false);
+  const [teacherId, setTeacherId] = useState<string>('');
+  const [isConverting, setIsConverting] = useState<boolean>(false);
+  const [conversionError, setConversionError] = useState<string | null>(null);
+
+  // Keep original uploaded data in state
+  const [originalCsvData, setOriginalCsvData] = useState<OriginalData[]>();
+  // Keep processed data ready for deletion
+  const [processedCsvData, setProcessedCsvData] = useState<ProcessedData[]>();
+
+  const handleFileUpload = async (
+    event: React.ChangeEvent<HTMLInputElement>
+  ) => {
+    const file = event.target.files?.[0];
+    setCsvError(null);
+
+    if (file && file.type === 'text/csv') {
+      setCsvFile(file);
+
+      try {
+        // Parse and validate CSV
+        const parsedData = await parseCsv(file);
+        setOriginalCsvData(parsedData);
+        setCsvData(parsedData);
+
+        // Determine if conversion is needed
+        if (hasUsernameColumn(parsedData)) {
```

Comment:
```suggestion
        if (hasUsernameColumn(parsedData) && !hasUserIds) {
```
nit: what if a csv has userIds and usernames? Conversion should work fine in that case, but we wouldn't necessary need to convert before.

Reply:
Correct... however, I don't see a scenario where we have userIds already as they are so opaque to users.